plugins/permission-node/src/ServerPermissionClient.ts

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { TokenManager, ServerTokenManager } from '@backstage/backend-common';
+import { Config } from '@backstage/config';
+import {
+  AuthorizeRequest,
+  AuthorizeRequestOptions,
+  AuthorizeResponse,
+  AuthorizeResult,
+  DiscoveryApi,
+  PermissionClient,
+  PermissionClientInterface,
+} from '@backstage/plugin-permission-common';
+
+/**
+ * A thin wrapper around
+ * {@link @backstage/plugin-permission-common#PermissionClient} that allows all
+ * backend-to-backend requests.
+ * @public
+ */
+export class ServerPermissionClient implements PermissionClientInterface {
+  private readonly serverTokenManager: TokenManager;
+  private readonly permissionClient: PermissionClient;
+  private readonly permissionEnabled: boolean;
+
+  constructor(options: {
+    discoveryApi: DiscoveryApi;
+    configApi: Config;
+    serverTokenManager: TokenManager;
+  }) {
+    const { discoveryApi, configApi, serverTokenManager } = options;
+    this.permissionClient = new PermissionClient({ discoveryApi, configApi });
```

Comment:
If we create a static factory method, the arguments to that method have to change for the same reasons, no? Unless we ask for the permissionClient to be passed into the static factory method as well, in which case the calling code might look a bit weird:
```
const permissionClient = new PermissionClient({ ... });
const serverPermissionClient = ServerPermissionClient.create({ permissionClient, ... });
```
If this looks ok, I'm happy to go that route!

Reply:
Yep a static `.create` would be good here. It's fine to construct a `PermissionClient` in there though, that's np. The thing to avoid is any kind of logic or setup in the constructor. It's best to strive for having constructors just assign values to members and nothing else. The most immediate and clear benefit of this imo is that you get a lot more flexibility in the API. If we want to change how this is constructed right now it's really tricky. If it's a static `.create` method, and especially if the constructor is also private, we're completely free to add other factory methods, builders, or what have you, along with deprecating existing ones.

Reply:
Ah, good point about the flexibility of the API for future deprecations & factory methods.

Reply:
[Done](https://github.com/backstage/backstage/pull/8357/commits/3a061cedd77b6786d0057c6410376fddb8e8c34e)!